market_maker_keeper/uniswapv2_market_maker_keeper.py

```diff
@@ -0,0 +1,379 @@
+
+        parser.add_argument("--initial-delay", type=int, default=10,
+                            help="Initial number of seconds to wait before placing liquidity")
+
+        parser.add_argument("--debug", dest='debug', action='store_true',
+                            help="Enable debug output")
+
+        self.arguments = parser.parse_args(args)
+        setup_logging(self.arguments)
+
+        self.web3 = kwargs['web3'] if 'web3' in kwargs else Web3(HTTPProvider(endpoint_uri=f"{self.arguments.rpc_host}:{self.arguments.rpc_port}",
+                                                                              request_kwargs={"timeout": self.arguments.rpc_timeout}))
+
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        if 'web3' not in kwargs:
+            register_keys(self.web3, self.arguments.eth_key)
+
+        # TODO: Add a more sophisticated regex for different variants of eth on the exchange 
+        # Record if eth is in pair, so can check which liquidity method needs to be used
+        self.is_eth = 'ETH' in self.pair()
+
+        # Assume token a is always ETH if ETH is in pair to simplify logic
+        token_a_name = 'WETH' if self.is_eth else self.pair().split('-')[0]
+        token_b_name = self.pair().split('-')[1]
+
+        self.reloadable_config = ReloadableConfig(self.arguments.token_config)
+        self._last_config_dict = None
+        self._last_config = None
+        token_config = self.get_token_config().tokens
+        self.token_a = list(filter(lambda token: token.name == token_a_name, token_config))[0]
+        self.token_b = list(filter(lambda token: token.name == token_b_name, token_config))[0]
+
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+
+        self.accepted_slippage = Wad.from_number(self.arguments.accepted_slippage / 100)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.initial_exchange_rate = self.arguments.initial_exchange_rate
+        self.uniswap_current_exchange_price = Wad.from_number(self.initial_exchange_rate) if self.initial_exchange_rate is not None else self.uniswap.get_exchange_rate()
+        self.testing_feed_price = False
+        self.test_price = Wad.from_number(0)
+
+        # used for local testing
+        if self.arguments.factory_address is None and self.arguments.router_address is None:
+            self.uniswap = UniswapV2(self.web3, self.token_a, self.token_b)
+        else:
+            self.uniswap = UniswapV2(self.web3, self.token_a, self.token_b, Address(self.arguments.router_address), Address(self.arguments.factory_address))
+
+    def main(self):
+        with Lifecycle(self.web3) as lifecycle:
+            lifecycle.initial_delay(self.arguments.initial_delay)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(10, self.place_liquidity)
+
+    def startup(self):
+        self.uniswap.approve(self.token_a)
```

Comment:
The PR with `pyexchange` implmentation (https://github.com/makerdao/pyexchange/pull/151/files#diff-8f7634f1e582be293e2d361a03d968caR117) uses `directly` for the two approval calls.  It'd be nice if an onchain call was added to check whether the approval needs to be performed before sending a TX.  Would save some gas each time the keeper was started.

Reply:
So, I can add comments around the approval, but that `directly` approve function will actually already check the allowance for a given token and contract prior to sending the approve transaction again unnecessarily: https://github.com/makerdao/pymaker/blob/master/pymaker/approval.py#L38